plugins/org.eclipse.n4js.generator/src/org/eclipse/n4js/generator/N4JSCompositeGenerator.java

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright (c) 2016 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.generator;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.log4j.Logger;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.n4js.N4JSGlobals;
+import org.eclipse.n4js.external.ExternalLibraryUriHelper;
+import org.eclipse.n4js.generator.common.CompilerDescriptor;
+import org.eclipse.n4js.generator.common.IComposedGenerator;
+import org.eclipse.n4js.generator.common.ISubGenerator;
+import org.eclipse.n4js.generator.common.SubgeneratorsRegistry;
+import org.eclipse.n4js.projectModel.IN4JSCore;
+import org.eclipse.n4js.projectModel.IN4JSSourceContainer;
+import org.eclipse.xtext.generator.IFileSystemAccess;
+
+import com.google.inject.Inject;
+
+/**
+ * A composite generator is responsible for a language. A composite generator delegates the generation logics to its
+ * children (subgenerators).
+ */
+public class N4JSCompositeGenerator implements IComposedGenerator {
+
+	private static final Logger LOGGER = Logger.getLogger(N4JSCompositeGenerator.class);
+
+	@Inject
+	private IN4JSCore n4jsCore;
+
+	@Inject
+	private ExternalLibraryUriHelper externalLibraryUriHelper;
+
+	@Inject
+	private SubgeneratorsRegistry subgeneratorRegistry;
+
+	@Override
+	public void doGenerate(Resource input, IFileSystemAccess fsa) {
+		if (isApplicableTo(input)) {
+			// Delegate to subgenerators only when the composite generator is applicable to the resource
+			for (ISubGenerator subgenerator : getSubGenerators()) {
+				subgenerator.doGenerate(input, fsa);
+			}
+		}
+	}
+
+	/*
+	 * (non-Javadoc)
+	 *
+	 * @see org.eclipse.n4js.generator.IComposedGenerator#getCompilerDescriptors()
+	 */
+	@Override
+	public Set<CompilerDescriptor> getCompilerDescriptors() {
+		Set<CompilerDescriptor> descriptors = new HashSet<>();
+		getSubGenerators().stream().forEachOrdered(subGenerator -> {
+			descriptors.add(subGenerator.getCompilerDescriptor());
+		});
+		return descriptors;
+	}
+
+	@Override
+	public boolean isApplicableTo(Resource input) {
+		if (N4JSGlobals.N4JS_FILE_EXTENSION.equals(input.getURI().fileExtension())
```

Comment:
1. Are `*.n4js.xt` files skipped on purpose?
2. Use `ResourceType` ? See `org.eclipse.n4js.generator.common.AbstractSubGenerator.hasValidFileExtension(Resource)`

Reply:
Unfortunately, `ResourceType` does not know N4JSX. This will change after https://github.com/eclipse/n4js/issues/368

In the future, we will need to refactor the whole generator infrastructure as described in https://github.com/eclipse/n4js/issues/382 anyway. So I would suggest, we will consider .xt file extensions in that task.